src/librustc_typeck/check/mod.rs

```diff
@@ -3408,6 +3408,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             return self.tcx.types.err;
         };
 
+        if self.tcx.sess.features.borrow().non_exhaustive {
+            let is_local = variant.did.is_local();
+
+            if !is_local {
```

Comment:
Looks like the check for `non_exhaustive` is forgotten here.

Reply:
Also see https://github.com/rust-lang/rust/pull/45394/files#r146130454, `non_exhaustive` on `enum` should not prohibit struct expressions using that enum's variants.

Reply:
Added check.